src/critics/prefer_forward_critic.cpp

```diff
@@ -0,0 +1,47 @@
+// Copyright 2022 FastSense, Samsung Research
+
+#include "mppic/critics/prefer_forward_critic.hpp"
+
+namespace mppi::critics
+{
+
+void PreferForwardCritic::initialize()
+{
+  auto node = parent_.lock();
+
+  auto getParam = utils::getParamGetter(node, name_);
+  getParam(power_, "prefer_forward_cost_power", 1);
+  getParam(weight_, "prefer_forward_cost_weight", 20.0);
+
+  RCLCPP_INFO(
+    logger_, "PreferForwardCritic instantiated with %d power and %f weight.", power_, weight_);
+}
+
+void PreferForwardCritic::score(
+  const geometry_msgs::msg::PoseStamped & /* robot_pos */, const xt::xtensor<double, 3> & trajectories,
+  const xt::xtensor<double, 2> & /* path */, xt::xtensor<double, 1> & costs,
+  nav2_core::GoalChecker * /* goal_checker */)
+{
+  using namespace xt::placeholders;  // NOLINT
+
+  xt::xtensor<T, 2> x_diff = xt::view(trajectories, xt::all(), xt::range(1, _), 0) -
+              xt::view(trajectories, xt::all(), xt::range(_, -1), 0);
+  xt::xtensor<T, 2> y_diff = xt::view(trajectories, xt::all(), xt::range(1, _), 1) -
+              xt::view(trajectories, xt::all(), xt::range(_, -1), 1);
+
+  xt::xtensor<T, 2> yaws = xt::view(trajectories, xt::all(), xt::range(_, -1), 2);
```

Comment:
More on my side for understanding:

The `x_diff` and `y_diff` are finding the delta X and Y's applied by the trajectories. I believe these in the reference frame of the map or the reference from of the robot location (e.g. is `(0,0)` the robot base)? 

If so, can't you just look at the sign of `x_diff` to determine if its in reverse?

I'm not sure I understand why you subtract `M_PI` from `xt::atan2(y_diff, x_diff) - yaws`. In fact, I'm not entirely sure why the actual angles `yaws` are important here if we're talking about translational reversing. 

The answer to that might also answer why the term is `xt::cos(thetas) * xt::sqrt(x_diff * x_diff + y_diff * y_diff);` instead of `xt::cos(thetas) * x_diff + xt::sin(theta) * y_diff` which would be the component of each in the `x` direction if the data is in the map frame?

Reply:
I don't think we should punish any lateral motions -- only the distance or speed at which we reverse. lateral motions could be from an omni platform that is ambiguous if its good or bad (since +lateral is left and -lateral is right)

Reply:
> The x_diff and y_diff are finding the delta X and Y's applied by the trajectories. I believe these in the reference frame of the map or the reference from of the robot location (e.g. is (0,0) the robot base)?

it's in reference frame of the map

> I'm not sure I understand why you subtract M_PI from xt::atan2(y_diff, x_diff) - yaws. In fact, I'm not entirely sure why the actual angles yaws are important here if we're talking about translational reversing.

As far as i know it takes opposite direction of an angle to let negative direction have plus sign. I think here it cloud be just reversing the sign

> The answer to that might also answer why the term is xt::cos(thetas) * xt::sqrt(x_diff * x_diff + y_diff * y_diff); instead of xt::cos(thetas) * x_diff + xt::sin(theta) * y_diff which would be the component of each in the x direction if the data is in the map frame?

Aren't these two expressions equivalent?

Reply:
> I don't think we should punish any lateral motions -- only the distance or speed at which we reverse. lateral motions could be from an omni platform that is ambiguous if its good or bad (since +lateral is left and -lateral is right) or for differential representing backing up and turning -- which I don't believe needs to be penalized more for non-straight backing up maneuvers.

Given as it is now it penalize only negative direction of translation projection on base_link forward axis

Reply:
No,

`math.cos(0.5) * 5.5 + math.sin(0.5) * 3.2` = 6.3608658139305
`math.cos(0.5) * math.sqrt(5.5*5.5 + 3.2*3.2)` = 5.584211688528694

Thus `cos(theta) * sqrt(dx*dx + dy*dy) != cos(theta) * dx + sin(theta) * dy`

Reply:
> As far as i know it takes opposite direction of an angle to let negative direction have plus sign. 

:+1:

Reply:
I'm not sure this formulation I understand. The `v_negative = xt::cos(thetas) * xt::sqrt(x_diff * x_diff + y_diff * y_diff);` I don't think actually works and I'm not sure actually what `yaw` has to do with this. 

It also ` xt::view(trajectories, xt::all(), xt::range(_, -1), 0);` is that the last point in the trajectory or the first? The `-1` makes me think last, which seems odd. I would think that a bunch of change in X and Ys would want to subtract the first point in each trajectory from the poses to find the total distance traversed in the trajectory and it would start with the integrated distance being `0` at the robot pose center

Reply:
> I'm not sure this formulation I understand. The v_negative = xt::cos(thetas) * xt::sqrt(x_diff * x_diff + y_diff * y_diff); I don't think actually works and I'm not sure actually what yaw has to do with this.

theta is angle between current primitive angle and direction of translation, so cos of this angle times translation gives projection on forward axis i guess (that was explained so to me by implementor of this) 
I'll put more thoughts on this now 

> It also xt::view(trajectories, xt::all(), xt::range(_, -1), 0); is that the last point in the trajectory or the first? The -1 makes me think last, which seems odd

first argument: is tensor - "trajectories", 
second argument: "xt::all()" - take all trajectories, 
third: "xt::range(_, -1)" - take time steps from the beginning to the end minus one (all except last)
fourth : "0" take x-es

Reply:
Maybe then diff_x/diff_y aren't what I thought they were. I thought they were a normalized set of trajectory points. If you have the robot at `robot_pose` and a set of `trajectories` both in map frame, each point in `trajectories` minus the `robot_pose` gives you the dx's and dy's of the trajectories relative to the robot base position. From that, you can easily tell which are going in the reverse direction and penalize them.

Reading this again, the dx/dys are subtracting their neighbors to find the incremental change per-point in the trajectories. 

I'm still not entirely clear on the `yaws`. If you know the robot's starting angle, shouldn't you use that to tell the angle relative to the robot position is "backwards"? That would be the wrong thing to do if you're trying to find each individual segment in each trajectory that goes backwards relative to its previous neighbor, which maybe is what you're doing here.

 I don't understand why the angle `thetas` matters behind the fact that its in a different direction to score it negatively if it goes in reverse. I also don't see where we integrate the costs of reversing across a trajectory to assign a score proportional to the amount it goes backwards. This looks like it takes the maximum of only 1 set of dx/dys backward motions between trajectory points. 

I still don't think `-xt::cos(thetas) * xt::hypot(dx, dy)` is correct. Everything is in map frame, so the component of `dx` in the "x" direction of the robot's pose is not necessarily the full component. As well with Y and being completely independent.

Reply:
yaws in map frame, while`thetas` is in base_link frame. `thetas` matters because it translated to local coordinate system yaws. That's why  `cos(theta) * dx + sin(theta) * dy` won't work out.
You have translation in one coordinate system and angle in another. I see here as a working solution just get absolute length of translation and then get cosine component